### ISSUE TITLE
Node features

### ISSUE THREAD

**Issue by jafarjamal** [deezer/gravity_graph_autoencoders#5]
Hello,

I would like to start by thanking you for this amazing work and very interesting paper. I have a very simple question. How do I define node features, and how do I load them inside 'input_data.py' instead of using and identity matrix?

Thanks.

**Comment by GuillaumeSalhaGalvan** [comment]
Dear @jafarjamal,

First of all, thank you very much for your feedback!

Let us assume that, in addition to the graph structure summarized by the adjacency matrix `adj`, you indeed have access to node-level feature vectors, say of dimension `f`, that bring additional information on each of your `n` nodes. Let us assume that you stacked and stored these node features  in the `data` folder in a `nodefeatures.txt` file with `n` lines and `f` columns. Then, you can load them in `input_data.py` by replacing the line:

```features = sp.identity(adj.shape[0])```

by:

```features = sp.coo_matrix(np.loadtxt("../data/nodefeatures/txt", delimiter='\t'))```

where, here, I use the COO sparse format for consistency w.r.t. the next preprocessing steps from the code. Specify your actual delimiter and path. After modification of `input_data.py`, do not forget to run again: 

```
cd ..
python setup.py install
```

and to specify `--features=True` within your options when running `train.py`. Then, your node features will be used as the initial embedding of the GCN encoder (see section 2.2. of the paper) during training.

**Comment by GuillaumeSalhaGalvan** [comment]
By the way, you might also want to take a look at [this repo](https://github.com/deezer/linear_graph_autoencoders) on graph AE/VAE, where we used node features to learn embeddings  for the ***un***directed versions of the Cora, Citeseer and Pubmed citation networks. Node features correspond to bag-of-word vectors associated to each node (which are scientific papers).

For these graphs, data loading* is a bit more complex than my previous message, due to the data format used and to some required node reordering**. But the idea is the same.

Best,

Guillaume



*These lines of code are borrowed from [the original GAE repo](https://github.com/tkipf/gae) from T. Kipf.

**If you use your own data, I'd advice you to always double-check that feature vectors are ordered in a consistent way w.r.t. how nodes from the graph are read/ordered.

**Comment by jafarjamal** [comment]
Thank you so much for your reply. That is exactly what i tried. I defined the node features as an NxN matrix (each node has N features) and loaded the data into a coo_matrix. However, when i enable node features and and start the training procedure, the training loss value becomes NaN. Sometimes the first epoch works, but after that it becomes NaN. 

I looked into one of the closed issues and changing increasing epsilon value might solve the problem. I tried several values but the problem remains the same, so it might not be related to the epsilon value.

Best,
Jafar

**Comment by GuillaumeSalhaGalvan** [comment]
Hi again @jafarjamal. Although it is a bit hard for me to provide help on this specific problem (as I do not know your data/graph/setup), I suspect it might be a data-related issue. I would:
- check that the feature vectors are normalized, and try normalizing them otherwise
- check whether, at least, the model runs without the features (i.e. verify that the problem is indeed caused by adding the features)
- verify whether the features vectors are correctly read and preprocessed in the lines of train.py before training.

Regarding the problem mentioned in #4, it is linked to the (non-variational) gravity-inspired graph AE model. Do you face the same NaN problem with the standard graph AE/VAE models, and/or with the gravity-inspired VAE model ? If yes, then it's something else.
